### DISCUSSION TITLE
Setting up jdtls

### DISCUSSION THREAD

**Discussion by couryrr** [VonHeikemen/lsp-zero.nvim#76] [Q&A]
As a note I am on a Windows 11 machine and I have java installed and on the path.

I do a lot of java work and wanted to get neovim setup. I moved over to a project and opened the main java class. lsp-zero and mason did what I expected and asked to install jdtls and it looks like everything is installed.

![image](https://user-images.githubusercontent.com/2702992/208337084-f9904cdb-660d-4868-8216-deb108a903d7.png)

I was not getting any auto complete or import management though. As a test I added a script to check that all was running

```
-- configure an individual server
lsp.configure('jdtls', {
  flags = {
    debounce_text_changes = 150,
  },
  on_attach = function(client, bufnr)
    print('lsp server (jdtls) attached')
  end
})

lsp.configure('gopls', {
  flags = {
    debounce_text_changes = 150,
  },
  on_attach = function(client, bufnr)
    print('lsp server (gopls) attached')
  end
})
```

Java is not picking up the lsp server
![image](https://user-images.githubusercontent.com/2702992/208337612-dfaef536-4f9c-4d77-a2da-ed8dbe4ce0da.png)

go seems to just fine and works for me.
![image](https://user-images.githubusercontent.com/2702992/208337547-72c5cde5-02de-44eb-913e-de99e0b60150.png)

I am sure this is a configuration issue somewhere but I am not able to find where. Does anyone have `jdtls` running and can you point me in the direction of the proper configuration.

**Comment by VonHeikemen** [comment]
You can use the command `LspLog` to inspect the lsp servers log file, maybe there's more info there.

Now I don't work with java but is my understanding people use this plugin [nvim-jdtls](https://github.com/mfussenegger/nvim-jdtls) to handle the language server. If you do use it, make sure you are not calling `jdtls` using lsp-zero.

```lua
local lsp = require('lsp-zero')
lsp.preset('recommended')

-- by calling build_options here will make lsp-zero skip the setup for jdtls
lsp.build_options('jdtls', {})

lsp.setup()
```

You could then configure in another place like it says in `nvim-jdtls` documentation.

```lua
local lsp_opts = lsp.build_options('jdtls', {})
require('jdtls').start_or_attach({
  on_attach = lsp_opts.on_attach,
  capabilities = lsp_opts.capabilities,
 ---- rest of the options......
})
```

**Reply by couryrr** [reply]
I was going based off this doc https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#jdtls that what mason had with jdtls would be fine for my current needs. 

Checking the logs I am not getting anything from my current setup. There was a typo in one of my configurations previously that shows up but nothing for the current jdtls issue.

That does help with how the configuration could be overridden.

**Reply by VonHeikemen** [reply]
Does the command `LspInfo` gives any error for jdtls?

**Reply by couryrr** [reply]
The configuration and data dir do not exist for `cmd`. The slashes are also a bit off too. Windows does not seem to care about either when I run the command directly though.

```
Use [q] or [Esc] to quit the window
 
 Language client log: C:\Users\Coury\AppData\Local\nvim-data\lsp.log
 Detected filetype:   java
 
 0 client(s) attached to this buffer: 
 
 Other clients that match the filetype: java
 
 Config: jdtls
 	Refer to :h lspconfig-root-detection for help.
 	filetypes:         java
 	root directory:    Not found.
 	cmd:               C:\Users\Coury\AppData\Local\nvim-data\mason\bin\jdtls.CMD -configuration C:\Users\Coury/.cache/jdtls/config -data C:\Users\Coury/.cache/jdtls/workspace
 	cmd is executable: true
 	autostart:         true
 	custom handlers:   workspace/applyEdit, textDocument/codeAction, $/progress, language/status, textDocument/rename
 
 Configured servers list: gopls, eslint, sumneko_lua, jdtls
```

**Reply by couryrr** [reply]
Thank you for the help so far. As a test I removed lsp-zero, just to simplify my search, to see if I could get the config working for Mason direct. I noticed a few items I could change but they had no effect on the over all issue.

This is an issue with either windows and Mason or jdtls.

Just for completness the lsp config for Mason looks like (which I could reproduce in `lsp-zero`):

```
require'lspconfig'.gopls.setup{}
require'lspconfig'.jdtls.setup{
    cmd = {'jdtls-win.cmd', "-configuration", "C:\\Users\\Coury\\AppData\\Local\\nvim-data\\mason\\packages\\jdtls\\config_win\\config.ini"},
    filetypes = { 'java', 'not-java'}
}
```

Again go works java is a pain (it never wins with config).

**Reply by VonHeikemen** [reply]
I looks like it can't find the root directory. Do you have any of these in the root of your project?

```
        -- Single-module projects
        {
          'build.xml', -- Ant
          'pom.xml', -- Maven
          'settings.gradle', -- Gradle
          'settings.gradle.kts', -- Gradle
        },
        -- Multi-module projects
        { 'build.gradle', 'build.gradle.kts' },
```

**Reply by couryrr** [reply]
I have a `pom.xml` in this particular case. I set the root dir in the config and even the data dir (which is the workspace). Current Mason config is below. It is not great practice to have the project and workspace in the same place but I was just testing. 

It is really hard to tell what is going on though since there is next to no logs coming out of anywhere. I am starting to think jdtls is not starting at all.

This uses the windows specific scripts. I tried moving over to the python version but that does not seem to run in Mason.

```
}
require'lspconfig'.jdtls.setup{
 cmd = {
        'jdtls-win.cmd',
        "-configuration",
        "C:\\Users\\Coury\\AppData\\Local\\nvim-data\\mason\\packages\\jdtls\\config_win",
        "-jar",
        "C:\\Users\\Coury\\AppData\\Local\\nvim-data\\mason\\packages\\jdtls\\plugins\\org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar",
        "-data",
        "C:\\Users\\Coury\\Documents\\Code\\interviews\\truleo\\app",
    },
    single_file_support = true,
    root_dir = function() 
        return "C:\\Users\\Coury\\Documents\\Code\\interviews\\truleo\\app"
    end,
    flags = {
        debounce_text_changes = 150,
    },
    on_attach = function(client, bufnr)
        print('lsp server (jdtls) attached')
    end
}
```

**Reply by VonHeikemen** [reply]
If with this config the server can't still start then is worth opening a issue in `nvim-lspconfig`'s repo.
